src/units/abilities.cpp

```diff
@@ -966,14 +989,222 @@ void attack_type::modified_attacks(bool is_backstab, unsigned & min_attacks,
+		if(apply_to.empty() || apply_to == "both" || apply_to == "self") {
+			return true;
+		}
+		if(attacker && apply_to == "attacker") {
+			return true;
+		}
+		if(!attacker && apply_to == "defender") {
+			return true;
+		}
+		++i;
+	}
+	return false;
+}
+
+bool leadership_affects_opponent(const std::string& ability, const unit &un, const map_location& loc, bool attacker)
+{
+	unit_ability_list abil = un.get_abilities(ability, loc);
+	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+		const std::string& apply_to = (*i->first)["apply_to"];
+		if(apply_to == "both" || apply_to == "opponent") {
+			return true;
+		}
+		if(attacker && apply_to == "defender") {
+			return true;
+		}
+		if(!attacker && apply_to == "attacker") {
+			return true;
+		}
+		++i;
+	}
+	return false;
+}
+
+unit_ability_list list_leadership(const std::string& ability, unit_const_ptr un, unit_const_ptr up, const map_location& loc, const map_location& opp_loc, bool attacker, const_attack_ptr weapon, const_attack_ptr opp_weapon)
+{
+	unit_ability_list abiln (loc);
+	unit_ability_list abil = (*un).get_abilities(ability, loc);
+	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+		const config &filter = (*i->first).child("filter_opponent");
+		const config &filter_attacker = (*i->first).child("filter_attacker");
+		const config &filter_defender = (*i->first).child("filter_defender");
+		bool fighter_filter = ability_apply_filter(un, up, ability, *i->first, loc, opp_loc, attacker, weapon, opp_weapon);
+		const std::string& active_on = (*i->first)["active_on"];
+		bool active_on_bool = !(active_on.empty() || (attacker && active_on == "offense") || (!attacker && active_on == "defense"));
+		bool weapon_filter = !((*un).ability_affects_weapon(*i->first, weapon, false) && (*un).ability_affects_weapon(*i->first, opp_weapon, true));
+		if(!up &&(filter || (filter_attacker && !attacker) || (filter_defender && attacker))) {
+			return abiln;
+		}
+
+		if(active_on_bool || fighter_filter || weapon_filter) {
+			i = abil.erase(i);
+		} else {
+			++i;
+		}
+	}
+	return abil;
+}
+
+unit_ability_list attack_type::list_ability(const std::string& ability) const
+{
+	const unit_map& units = display::get_singleton()->get_units();
+
+	unit_const_ptr self = self_;
+	unit_const_ptr other = other_;
+
+	if(self == nullptr) {
+		unit_map::const_iterator it = units.find(self_loc_);
+		if(it.valid()) {
+			self = it.get_shared_ptr().get();
+		}
+	}
+	if(other == nullptr) {
+		unit_map::const_iterator it = units.find(other_loc_);
+		if(it.valid()) {
+			other = it.get_shared_ptr().get();
+		}
+	}
+
+	// Make sure they're facing each other.
+	temporary_facing self_facing(self, self_loc_.get_relative_dir(other_loc_));
+	temporary_facing other_facing(other, other_loc_.get_relative_dir(self_loc_));
+	unit_ability_list abil_list(self_loc_);
+	if(self){
+		if(leadership_affects_self(ability, *self, self_loc_, is_attacker_)) {
```

Comment:
i wonder why this check here is needd maybe you can explain?

Reply:
@gfgtdf This check is necessary for what avoid bug in attack_prediction windows with unit.facing.

Reply:
:/ i think there shodul be a better fix for this one, can you explaing the bug that happen otherwise or link to a bugreport if one exists?

Reply:
@gfgtdf if you speak of if(self), it is because of travis validation fail before what i put this

Reply:
@gfgtdf the bug for unit.facingis reported here https://github.com/wesnoth/wesnoth/issues/4475
and temporary_facing other_facing(other, other_loc_.get_relative_dir(self_loc_)); is used for specials for avoid this bugs with backstab

Reply:
noo i was talking about the `if(leadership_affects_self(ability, *self, self_loc_, is_attacker_ ... ` line not the temporary_facing object.

Reply:
like, wouldn't list_leadership returns an empty list anyway if leadership_affects_self returned false? So why do we need to check for leadership_affects_self before

Reply:
if leadership_affects_self is false then ability not applied to student, and 
			abil_list = list_leadership(ability, self, other, self_loc_, other_loc_, is_attacker_, shared_from_this(), other_attack_); don't must be called, else student will be affected same if apply_to=opponent.

Reply:
but list_leadership doe also checks for apply_to=opponent/self

Reply:
@gfgtdf no, list_leadership do check for active_on and filter_student,opponent,etc, but not for apply_to=opponent/self

Reply:
hmm but then wouldnt it wtill be wrong? leadership_affects_self just cehcks whether the unit has any ability the the correct apply_to, but a unit might have multiple abilities and its possibel that only  one  of them might be apply_to=self and ther other apply_to=opponent, and with your code it looks like it'd return both of them if they'd for example look for the leadership that applies to self.

Reply:
@gfgtdf i tested marksman abilitie without  leadership_affects_self with apply_to=opponent and the two unit are affected with cth=60% in forest instead of 30%

Reply:
well yes that's the code is wrong without that line, doesn't mean that the code is correct with that line. Whjat im saying is that from how i read the code it looks like a unit with an ability like
```
[damange]
  add=10
  apply_to=self
[/damange]
[damange]
  add=20
  apply_to=opponent
[/damange]
```
would add 30 damange to both attacker and defender becasue in both cases leadership_affects_self returns true, and you said list_leadership ignores apply_to

Reply:
@gfgtdf oh no! this is the highest value who is returned! and i don't see how fix that!